File: ios/PacketTunnel/StartupPathVerification.swift
import Foundation

let startupStageDnsProbe = "dns_probe"
let startupStageInternetProbe = "internet_probe"

private enum VerifiedStartupPathEvent {
    case probe(InternetProbeResult)
    case engineStopped(String?)
}

/// Runs the startup verification (fresh DNS, then HTTPS, both through the tunnel) raced against
/// the engine's stop signal, and projects a failure into the fallback taxonomy:
///
///  - engine stop / local evidence → `LocalTunnelError` (aborts the whole relay ladder);
///  - DNS-stage remote failure → stage `dns_probe`. The emitted resolver chain already fell
///    over in-engine, so this stage failing means NO configured resolver answered through this
///    transport — evidence against the path, not one resolver;
///  - HTTPS-stage remote failure → stage `internet_probe`;
///  - remote failures type as `WssTransportError` when `wssFrontID` is set (WSS transport),
///    else `DirectPathError` — the only type that can unlock WSS fallback.
///
/// The engine is threaded in as closures so the seam stays hostlessly testable (the concrete
/// engine imports Libbox). `prepareForExpectedStop` linearizes remote classification against
/// libbox's stop callback exactly as before the extraction: if the callback already won, this is
/// a local engine failure; if teardown wins, a later callback is expected because the failed
/// candidate is about to be replaced. Extracted from the provider so startup truth-telling is
/// hostlessly testable: CONNECTED may only be published when this returns.
func verifyStartupTunnelPath(
    probe: @escaping @Sendable () async throws -> InternetProbeResult,
    waitForUnexpectedStop: @escaping @Sendable () async -> String?,
    hasUnexpectedStop: () -> Bool,
    prepareForExpectedStop: () -> Bool,
    wssFrontID: String?
) async throws -> InternetProbeResult {
    do {
        let result = try await withThrowingTaskGroup(of: VerifiedStartupPathEvent.self) { group in
            group.addTask { .probe(try await probe()) }
            group.addTask { .engineStopped(await waitForUnexpectedStop()) }
            defer { group.cancelAll() }
            while let event = try await group.next() {
                switch event {
                case .probe(let result):
                    return result
                case .engineStopped(let reason):
                    if let reason {
                        throw LocalTunnelError(
                            stage: "active_tunnel_engine",
                            underlying: PacketTunnelProxyEngineError.engineStartFailed(reason)
                        )
                    }
                    try Task.checkCancellation()
                }
            }
            throw CancellationError()
        }
        guard hasUnexpectedStop() == false else {
            throw LocalTunnelError(
                stage: "active_tunnel_engine",
                underlying: PacketTunnelProxyEngineError.engineStartFailed(
                    "libbox stopped during startup verification"
                )
            )
        }
        return result
    } catch is CancellationError {
        throw CancellationError()
    } catch {
        if hasUnexpectedStop() {
            throw LocalTunnelError(
                stage: "active_tunnel_engine",
                underlying: PacketTunnelProxyEngineError.engineStartFailed(
                    "libbox stopped during startup verification"
                )
            )
        }
        let stage = error is DnsPathUnverifiedError ? startupStageDnsProbe : startupStageInternetProbe
        guard isGenuineRemoteDataPathFailure(error) else {
            throw LocalTunnelError(stage: stage, underlying: error)
        }
        guard prepareForExpectedStop() else {
            throw LocalTunnelError(
                stage: "active_tunnel_engine",
                underlying: PacketTunnelProxyEngineError.engineStartFailed(
                    "libbox stopped while classifying the startup path failure"
                )
            )
        }
        if let wssFrontID {
            throw WssTransportError(stage: stage, frontID: wssFrontID, underlying: error)
        }
        throw DirectPathError(stage: stage, underlying: error)
    }
}

File: android/app/src/main/java/com/openrung/vpn/StartupPathVerification.kt
package com.openrung.vpn

import com.openrung.net.DnsPathUnverifiedException
import com.openrung.net.InternetProbeResult
import kotlinx.coroutines.CancellationException

internal const val STARTUP_STAGE_DNS_PROBE = "dns_probe"
internal const val STARTUP_STAGE_INTERNET_PROBE = "internet_probe"

/**
 * Runs the startup verification (fresh DNS, then HTTPS, both through the tunnel) raced against
 * the engine's stop signal, and projects a failure into the fallback taxonomy:
 *
 *  - engine stop / local evidence → [LocalTunnelException] (aborts the whole relay ladder);
 *  - DNS-stage remote failure → stage `dns_probe`, after [onDnsPathFailure] has run so the
 *    resolver rotation advances before any retry builds its configuration;
 *  - HTTPS-stage remote failure → stage `internet_probe`;
 *  - remote failures type as [WssTransportException] when [wssFrontId] is set (WSS transport),
 *    else [DirectPathException] — the only type that can unlock WSS fallback.
 *
 * Extracted from the service so startup truth-telling is hostlessly testable: CONNECTED may only
 * be published when this returns.
 */
internal suspend fun verifyStartupTunnelPath(
    probe: suspend () -> InternetProbeResult,
    awaitUnexpectedEngineStop: suspend () -> String,
    wssFrontId: String?,
    onDnsPathFailure: () -> Unit,
): InternetProbeResult = try {
    awaitStartupProbeOrEngineStop(probe, awaitUnexpectedEngineStop)
} catch (error: CancellationException) {
    throw error
} catch (error: Throwable) {
    if (error is LocalTunnelException) throw error
    val dnsStage = error is DnsPathUnverifiedException
    val stage = if (dnsStage) STARTUP_STAGE_DNS_PROBE else STARTUP_STAGE_INTERNET_PROBE
    if (!isGenuineRemoteDataPathFailure(error)) {
        throw LocalTunnelException(stage, error)
    }
    // Only a failure classified as REMOTE is evidence about the resolver path; rotating on
    // local evidence would burn a healthy resolver and emit false failover telemetry. Still
    // ahead of the typed throws, so any retry below builds its config from the rotated order.
    if (dnsStage) onDnsPathFailure()
    if (wssFrontId != null) {
        throw WssTransportException(stage, wssFrontId, error)
    }
    throw DirectPathException(stage, error)
}
